File: internal/gira/templates.go
package gira

import (
	"fmt"
	"io/fs"
	"sort"
	"strings"

	templatefs "github.com/StatPan/gira/templates"
)

type RenderedTemplate struct {
	Path    string
	Content string
}

func RenderTemplateTree(template string, repo RepoRef, createdAt string) ([]RenderedTemplate, error) {
	if strings.Contains(template, "/") || strings.Contains(template, "\\") || template == "" {
		return nil, fmt.Errorf("invalid template name: %s", template)
	}
	if template != "default" {
		return nil, fmt.Errorf("unknown template: %s", template)
	}

	root := "default"
	var paths []string
	if err := fs.WalkDir(templatefs.FS, root, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if d.IsDir() {
			return nil
		}
		paths = append(paths, path)
		return nil
	}); err != nil {
		return nil, err
	}
	sort.Strings(paths)

	context := map[string]string{
		"repo_owner":     repo.Owner,
		"repo_name":      repo.Name,
		"repo_full_name": repo.FullName(),
		"created_at":     createdAt,
	}

	rendered := make([]RenderedTemplate, 0, len(paths))
	for _, path := range paths {
		contentBytes, err := templatefs.FS.ReadFile(path)
		if err != nil {
			return nil, err
		}
		rel := strings.TrimPrefix(path, root+"/")
		outRel := strings.TrimSuffix(rel, ".j2")
		content := renderSimpleTemplate(string(contentBytes), context)
		if strings.Contains(content, "{{") || strings.Contains(content, "{%") || strings.Contains(content, "{#") {
			return nil, fmt.Errorf("unsupported template expression in %s", rel)
		}
		rendered = append(rendered, RenderedTemplate{Path: outRel, Content: content})
	}

	return rendered, nil
}

func FormatDryRun(rendered []RenderedTemplate) string {
	var b strings.Builder
	for _, item := range rendered {
		b.WriteString("--- ")
		b.WriteString(item.Path)
		b.WriteString("\n")
		b.WriteString(strings.TrimRight(item.Content, "\n"))
		b.WriteString("\n\n")
	}
	return strings.TrimRight(b.String(), "\n") + "\n"
}

func renderSimpleTemplate(input string, context map[string]string) string {
	output := input
	for key, value := range context {
		output = strings.ReplaceAll(output, "{{ "+key+" }}", value)
		output = strings.ReplaceAll(output, "{{"+key+"}}", value)
	}
	return output
}

File: gira/templates.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Iterable

from jinja2 import Environment, FileSystemLoader, StrictUndefined

from gira.config import RepoRef


@dataclass(frozen=True)
class RenderedTemplate:
    path: str
    content: str


_TEMPLATE_ROOT = Path(__file__).resolve().parent.parent / "templates"


def render_template_tree(template: str, repo: RepoRef, created_at: str) -> list[RenderedTemplate]:
    """Render every file in a Gira template tree in deterministic path order."""
    if Path(template).name != template:
        raise ValueError(f"invalid template name: {template}")

    source = _TEMPLATE_ROOT / template
    if not source.is_dir():
        raise ValueError(f"unknown template: {template}")

    env = Environment(
        loader=FileSystemLoader(str(source)),
        undefined=StrictUndefined,
        autoescape=False,
        keep_trailing_newline=True,
    )
    context = {
        "repo_owner": repo.owner,
        "repo_name": repo.name,
        "repo_full_name": repo.full_name,
        "created_at": created_at,
    }

    rendered: list[RenderedTemplate] = []
    for path in _iter_template_files(source):
        rel = path.relative_to(source).as_posix()
        out_rel = rel[:-3] if rel.endswith(".j2") else rel
        content = env.get_template(rel).render(context)
        rendered.append(RenderedTemplate(path=out_rel, content=content))
    return rendered


def format_dry_run(rendered: Iterable[RenderedTemplate]) -> str:
    """Return a stable, human-readable dry-run rendering."""
    lines: list[str] = []
    for item in rendered:
        lines.append(f"--- {item.path}")
        lines.append(item.content.rstrip())
        lines.append("")
    return "\n".join(lines).rstrip() + "\n"


def _iter_template_files(source: Path) -> list[Path]:
    return sorted(path for path in source.rglob("*") if path.is_file())
